moquette/broker/src/main/java/io/moquette/broker/MQTTConnection.java

```diff
@@ -308,7 +306,7 @@ private boolean login(final ClientData clientData) {
 
     void handleConnectionLost() {
         String clientID = NettyUtils.clientID(channel);
-        if (clientID == null || clientID.isEmpty()) {
+        if (clientID == null || clientID.isEmpty() || bindedSession == null) {
```

Comment:
Why remove this?
Can we not get "Connection Lost" event if the bindedSession is null anymore?

Reply:
A larger refactor would be needed to get rid of this. The problem is for "case 3" where we disconnect a client and replace the MQTTConnection in the Session object - it screws up Session state.

What happens is that the Session transitions back in to the CONNECTED state and binds to newConnection. Then the `handleConnectionLost` for oldConnection triggers (in a separate thread) and transitions the Session to the DISCONNECTED state.

Reply:
Left a comment below about this

Reply:
Sorry, I was reading it slightly wrong. 
I get the intent now. Ya, a comment will be helpful.